scopus-import/src/main/java/no/sikt/nva/scopus/conversion/files/ScopusFileConverter.java

```diff
@@ -99,30 +105,57 @@ public ScopusFileConverter(HttpClient httpClient, S3Client s3Client, Environment
     }
 
     public List<AssociatedArtifact> fetchAssociatedArtifacts(DocTp docTp) {
-        var associatedArtifactsFromXmlReferences = extractAssociatedArtifactsFromFileReference(docTp);
+        var associatedArtifactsFromXmlReferences = extractAssociatedArtifactsFromFileReference(docTp)
+                                                       .stream()
+                                                       .filter(File.class::isInstance)
+                                                       .filter(this::isValid)
+                                                       .toList();
         return !associatedArtifactsFromXmlReferences.isEmpty() ? associatedArtifactsFromXmlReferences
                    : extractAssociatedArtifactsFromDoi(docTp);
     }
 
+    private boolean isValid(AssociatedArtifact associatedArtifact) {
+        return nonNull(((File) associatedArtifact).getMimeType());
+    }
+
     private static CrossrefResponse toCrossrefResponse(String body) throws JsonProcessingException {
         return JsonUtils.dtoObjectMapper.readValue(body, CrossrefResponse.class);
     }
 
     private static String getFilename(HttpResponse<InputStream> response) {
-        return extractContentDisposition(response).map(ScopusFileConverter::extractFileNameFromContentDisposition)
-                   .orElseGet(() -> randomUUID() + FILE_NAME_DELIMITER + PDF_FILE_TYPE);
+        return response.headers().firstValue(Headers.CONTENT_DISPOSITION)
+                   .map(ScopusFileConverter::extractFileNameFromContentDisposition)
+                   .or(() -> extractFileNameFromUrl(response))
+                   .or(() -> randomFileNameWithMimeTypeFromContentTypeHeader(response.headers()))
+                   .orElseGet(ScopusFileConverter::randomStringPdfFileName);
+    }
+
+    private static String randomStringPdfFileName() {
+        return fileNameWithExtension(PDF_FILE_TYPE);
     }
 
-    private static Optional<String> extractContentDisposition(HttpResponse<InputStream> response) {
-        var contentDisposition = response.headers().map().getOrDefault(Headers.CONTENT_DISPOSITION, List.of());
-        return contentDisposition.isEmpty() ? Optional.empty() : Optional.of(contentDisposition.get(0));
+    private static Optional<String> randomFileNameWithMimeTypeFromContentTypeHeader(HttpHeaders headers) {
+        return headers.firstValue(Header.CONTENT_TYPE)
+                   .map(ScopusFileConverter::extractFileExtensionFromContentType)
+                   .map(ScopusFileConverter::fileNameWithExtension);
     }
 
-    private static String getContentType(HttpResponse<InputStream> response) {
-        return Optional.of(response.headers().firstValue(Headers.CONTENT_TYPE))
-                   .map(optional -> optional.orElse(null))
-                   .map(value -> value.split(CONTENT_TYPE_DELIMITER)[0])
-                   .orElse(DEFAULT_CONTENT_TYPE);
+    @NotNull
```

Comment:
What does this jet brains annotation give us?

Reply:
They are auto generated, removing them